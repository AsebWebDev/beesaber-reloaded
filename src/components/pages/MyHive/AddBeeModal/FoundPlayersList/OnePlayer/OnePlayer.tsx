import { MDBIcon } from 'mdb-react-ui-kit';
import styled from 'styled-components';

import createLocalImageUrl from '@/helper/createLocalImageUrl';
import { parseAvatarUrl } from '@/helper/urlParser';

import type { PlayerInfo } from '@/sharedTypes/ScoreSaberUserInfo';

const Player = styled.div`
  display: flex;
`;

const Avatar = styled.img`
  margin-right: 1rem;
  max-width: 2rem;
`;

type Props = {
  handleSelect?: (user: PlayerInfo) => void;
  isAlreadyAdded: boolean;
  isOnlyResult: boolean;
  player: PlayerInfo;
};

function OnePlayer({
  handleSelect,
  isAlreadyAdded,
  isOnlyResult,
  player,
}: Props): JSX.Element | null {
  const { rank, country, playerId, playerName, avatar } = player;
  const url =
    avatar === '/images/steam.png' || avatar === '/images/oculus.png'
      ? createLocalImageUrl('bee.jpg')
      : parseAvatarUrl(avatar);

  return (
    <tr key={playerId}>
      <td>
        <Player>
          <Avatar src={url} alt={`Avatar of player ${playerName}`} />
          {playerName}
        </Player>
      </td>
      <td>{rank}</td>
      <td>{country}</td>
      <td>
        {handleSelect !== undefined && !isAlreadyAdded && (
          <MDBIcon
            data-testid="select-icon"
            onClick={() => handleSelect(player)}
            fas
            icon={isOnlyResult ? 'minus-circle' : 'plus-circle'}
          />
        )}
      </td>
    </tr>
  );
}

export default OnePlayer;
