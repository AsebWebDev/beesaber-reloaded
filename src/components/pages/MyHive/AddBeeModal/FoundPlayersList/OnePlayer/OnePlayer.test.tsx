import { render, screen } from '@testing-library/react';

import exampleSSUserInfo from '@/testing/testData/exampleSSUserInfo';

import OnePlayer from './OnePlayer';

const player = exampleSSUserInfo.playerInfo;

describe('OnePlayer', () => {
  it.each`
    amount        | isOnlyResult
    ${'one'}      | ${true}
    ${'multiple'} | ${false}
  `(
    'should match snapshot for $amount found player(s)',
    (isOnlyResult: boolean) => {
      render(
        <table>
          <tbody>
            <OnePlayer
              handleSelect={jest.fn()}
              player={player}
              isAlreadyAdded={false}
              isOnlyResult={isOnlyResult}
            />
          </tbody>
        </table>
      );

      const component = screen.getByRole('row');

      expect(component).toMatchSnapshot();
    }
  );

  it.each`
    amount        | isOnlyResult | icon
    ${'one'}      | ${true}      | ${'minus'}
    ${'multiple'} | ${false}     | ${'plus'}
  `(
    'should show $icon icon, when $amount player(s) found',
    ({
      icon,
      isOnlyResult,
    }: {
      icon: 'minus' | 'plus';
      isOnlyResult: boolean;
    }) => {
      render(
        <table>
          <tbody>
            <OnePlayer
              handleSelect={jest.fn()}
              player={player}
              isAlreadyAdded={false}
              isOnlyResult={isOnlyResult}
            />
          </tbody>
        </table>
      );

      const selectableIcon = screen.getByTestId('select-icon');

      expect(selectableIcon).toHaveClass(`fa-${icon}-circle`);
    }
  );
  it('should hide icon, when player is already added', () => {
    render(
      <table>
        <tbody>
          <OnePlayer
            handleSelect={jest.fn()}
            player={player}
            isAlreadyAdded={true}
            isOnlyResult={false}
          />
        </tbody>
      </table>
    );

    const lastCell = screen.getByRole('row').lastChild?.firstChild; // grabs last cell in row

    expect(lastCell).toBeNull();
  });

  it.each`
    avatar           | avatarLink
    ${'not present'} | ${undefined}
    ${'oculus'}      | ${'/images/oculus.png'}
    ${'steam'}       | ${'/images/steam.png'}
  `(
    'should show avatar placeholder, when avatar is $avatar',
    (avatarLink: string) => {
      const mockedPlayer = { ...player, avatar: avatarLink };

      render(
        <table>
          <tbody>
            <OnePlayer
              handleSelect={jest.fn()}
              player={mockedPlayer}
              isAlreadyAdded={false}
              isOnlyResult={false}
            />
          </tbody>
        </table>
      );

      const avatar = screen.getByRole('img', {
        name: `Avatar of player ${mockedPlayer.playerName}`,
      }) as HTMLImageElement;

      expect(avatar.src).toContain(`https://new.scoresaber.com${avatarLink}`);
    }
  );

  it('should call handleSelect, when icon clicked', () => {
    const mockHandleSelect = jest.fn();

    render(
      <table>
        <tbody>
          <OnePlayer
            handleSelect={mockHandleSelect}
            player={player}
            isAlreadyAdded={false}
            isOnlyResult={false}
          />
        </tbody>
      </table>
    );

    const selectableIcon = screen.getByTestId('select-icon');

    selectableIcon.click();

    expect(mockHandleSelect).toHaveBeenCalledTimes(1);
  });
});
