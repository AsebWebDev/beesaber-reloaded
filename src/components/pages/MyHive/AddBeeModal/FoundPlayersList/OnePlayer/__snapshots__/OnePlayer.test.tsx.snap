// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnePlayer should match snapshot for multiple found player(s) 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  margin-right: 1rem;
  max-width: 2rem;
}

<tr>
  <td>
    <div
      class="c0"
    >
      <img
        alt="Avatar of player cerret"
        class="c1"
        src="https://new.scoresaber.com/api/static/avatars/76561198333869741.jpg"
      />
      cerret
    </div>
  </td>
  <td>
    1
  </td>
  <td>
    US
  </td>
  <td>
    <i
      class="fas fa-minus-circle"
      data-testid="select-icon"
    />
  </td>
</tr>
`;

exports[`OnePlayer should match snapshot for one found player(s) 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  margin-right: 1rem;
  max-width: 2rem;
}

<tr>
  <td>
    <div
      class="c0"
    >
      <img
        alt="Avatar of player cerret"
        class="c1"
        src="https://new.scoresaber.com/api/static/avatars/76561198333869741.jpg"
      />
      cerret
    </div>
  </td>
  <td>
    1
  </td>
  <td>
    US
  </td>
  <td>
    <i
      class="fas fa-minus-circle"
      data-testid="select-icon"
    />
  </td>
</tr>
`;
